### PR TITLE
Make types compatible with both HTML and SVG elements

### PR DESCRIPTION
**Pull request by devongovett**
We use `HTMLAttributes<HTMLElement>` in a lot of return types from React Aria hooks. In addition, we often accept refs for `RefObject<HTMLElement>`. However, in many cases it is also possible to use React Aria with SVG elements, which support all of the ARIA attributes, and most of the events we use. For example, I made a Meter component using SVG. However, the current types wouldn't allow this if you use TypeScript.

This replaces most instances of `HTMLAttributes<HTMLElement>` with a new type: `DOMAttributes` exported from `@react-types/shared`. This type includes an intersection of attributes supported by both HTML and SVG elements.

In addition, most usages of `HTMLElement` (e.g. in refs and other APIs), have either been replaced with `Element`, or with a new type called `FocusableElement` which is an element plus a `focus()` method and `tabIndex` attribute (both supported by SVG as well).

There are a few places where we do rely on HTML element specific features (notably drag and drop), so those have been retained where necessary. However, mostly all components can be used with SVG elements with TypeScript now.

### CODE REVIEW

**Comment by majornista** [comment]
Yay, we've needed this for a while.

File: packages/@react-aria/interactions/src/textSelection.ts [hunk 1]
```diff
@@ -31,25 +31,25 @@ type State = 'default' | 'disabled' | 'restoring';
...
+export function disableTextSelection(target?: Element) {
   if (isIOS()) {
     if (state === 'default') {
       savedUserSelect = document.documentElement.style.webkitUserSelect;
       document.documentElement.style.webkitUserSelect = 'none';
     }
 
     state = 'disabled';
-  } else if (target) {
+  } else if (target instanceof HTMLElement || target instanceof SVGElement) {
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/textSelection.ts:44]: this was originally much wider, what is reason for the change?

File: packages/@react-aria/interactions/src/usePress.ts [hunk 2]
```diff
@@ -291,13 +291,13 @@ export function usePress(props: PressHookProps): PressResult {
...
         state.isPressed = false;
-        let target = e.target as HTMLElement;
+        let target = e.target as Element;
         triggerPressEnd(createEvent(state.target, e), 'keyboard', state.target.contains(target));
         removeAllGlobalListeners();
 
         // If the target is a link, trigger the click method to open the URL,
         // but defer triggering pressEnd until onClick event handler.
-        if (state.target.contains(target) && isHTMLAnchorLink(state.target) || state.target.getAttribute('role') === 'link') {
+        if (state.target instanceof HTMLElement && (state.target.contains(target) && isHTMLAnchorLink(state.target) || state.target.getAttribute('role') === 'link')) {
```
Comment (snowystinger) [inline packages/@react-aria/interactions/src/usePress.ts:300]: won't isHTMLAnchorLink be enough to prevent us from running this for an SVG? or why was this changed in this way? I'm not quite following

File: packages/@react-aria/slider/src/useSlider.ts [hunk 3]
```diff
@@ -74,7 +75,8 @@ export function useSlider(
       currentPosition.current = null;
     },
     onMove({deltaX, deltaY}) {
-      let size = isVertical ? trackRef.current.offsetHeight : trackRef.current.offsetWidth;
+      let {height, width} = trackRef.current.getBoundingClientRect();
```
Comment (snowystinger) [inline packages/@react-aria/slider/src/useSlider.ts:78]: for anyone else reviewing, offsetHeight does not exist on SVG's (it does it some browsers, but it shouldn't)

File: packages/@react-aria/slider/src/useSlider.ts [hunk 4]
```diff
@@ -168,7 +171,7 @@ export function useSlider(
     }
   };
 
-  if (labelProps.htmlFor) {
+  if ('htmlFor' in labelProps && labelProps.htmlFor) {
```
Comment (snowystinger) [inline packages/@react-aria/slider/src/useSlider.ts:174]: what is this change about? wouldn't `labelProps?.htmlFor` work?

File: packages/@react-aria/interactions/src/usePress.ts (hunk 2 above)
Comment (devongovett) [reply packages/@react-aria/interactions/src/usePress.ts:300]: It's because of the below line: only HTMLElements have a `click()` method.

File: packages/@react-aria/slider/src/useSlider.ts (hunk 4 above)
Comment (devongovett) [reply packages/@react-aria/slider/src/useSlider.ts:174]: The type returned by `useLabel` is `DOMAttributes | LabelHTMLAttributes<HTMLLabelElement>`, depending on the `labelElementType`. If the label is not a `<label>` element, then the type won't have an `htmlFor` prop, so TS requires this check.

File: packages/@react-aria/interactions/src/usePress.ts (hunk 2 above)
Comment (snowystinger) [reply packages/@react-aria/interactions/src/usePress.ts:300]: ah, typescript stikes again, thanks

**Review by snowystinger — APPROVED** [review]

**Review by reidbarber — APPROVED** [review]
